<!DOCTYPE html><html><head><title>HGRAINBOW</title><meta charset='utf-8'><link href='https://d31qtcb7c58z3n.cloudfront.net/res-min/themes/marxico.css' rel='stylesheet'><style>.note-content  {font-family: Helvetica, arial, freesans, clean, meiryo, sans-serifa;}


.note-content  {font-family: Helvetica, arial, freesans, clean, meiryo, sans-serifa;}
</style></head><body><div id='preview-contents' class='note-content'>
                        
                    

<p></p>

<h1 id="hgrainbow">HGRAINBOW</h1>

<h4 id="supplementary-data-and-scripts-used-in-the-article-haplotype-based-genome-wide-association-study-using-a-novel-snp-set-method-rainbow">Supplementary data and scripts used in the article “Haplotype-based genome wide association study using a novel SNP-set method : RAINBOW”</h4>

<h4 id="author-kosuke-hamazaki-hamazakiut-biometorg">Author : Kosuke Hamazaki (hamazaki@ut-biomet.org)</h4>

<h4 id="date-20190328">Date : 2019/03/28</h4>

<p>Here, I explain the structure of this repository.</p>

<hr>

<p><code>HGRAINBOW</code> <br>
├── <code>data.zip</code> : This folder contains datasets analyzed in this study. Please download and then decompress. <br>
│   ├── <code>extra</code> : This folder contain datasets other than genotype and phenotype. <br>
│   │   ├── <code>Table_S1_origin.csv</code> : The original data for <code>Table S1</code>. <br>
│   │   ├── <code>additive_relationship_matrix.RData</code> : The additive genetic relationship matrix used in this study. <br>
│   │   ├── <code>haplotype_block_list.csv</code> : The list of haplotype block information used in the SNP-set GWAS methods. <br>
│   │   └── <code>map2.csv</code> : The physical map corresponding to the haplotype block list. <br>
│   ├── <code>genotype</code> : The marker genotype used in this study. <br>
│   │   ├── <code>L3024_core_extract_L414_ind1A_ind1B_MAF_0.025_geno.tsv</code> : The marker genotype of 414 accessions used in this study. <br>
│   │   ├── <code>L3024_core_extract_L414_ind1A_ind1B_MAF_0.025_haplo1.tsv</code> : The marker haplotype of 414 accessions used in this study. <br>
│   │   └── <code>L3024_core_extract_L414_ind1A_ind1B_MAF_0.025_haplo2.tsv</code> : The marker haplotype of 414 accessions used in this study. <br>
│   └── <code>phenotype</code> : The phenotypic data simulated in this study. <br>
│       ├── <code>number_of_causals=2_direction_of_effect=minus_trial_no=1.csv</code> : The phenotypic data for the repulsion scenario. <br>
│       └── <code>number_of_causals=2_direction_of_effect=plus_trial_no=1.csv</code> : The phenotypic data for the coupling scenario. <br>
│       ├── <code>seeds_number_of_causals=2_direction_of_effect=minus_trial_no=1.csv</code> : The random seeds used for simulating phenotypic data for the repulsion scenario. <br>
│       └── <code>seeds_number_of_causals=2_direction_of_effect=plus_trial_no=1.csv</code> : The random seeds used for simulating phenotypic data for the coupling scenario. <br>
└── <code>scripts</code> : This folder contains scripts with the <code>R</code> language used in this study. <br>
    ├── <code>0.0_Rice_HGRAINBOW_subpop_list_to_generate_haplotype_data.R</code> : Extract subpopulation list from 3,000 accessions. <br>
    ├── <code>0.1_Rice_HGRAINBOW_haplotype_data.txt</code>  : Extract haplotype data of 414 accessions (<code>plink</code> and <code>vcftools</code>) adn estimate haplotype blocks by <code>plink</code> (not <code>R</code>!). <br>
    ├── <code>0.2_Rice_HGRAINBOW_modyfing_haplotype_block_list.R</code> : Modify haplotype block data estimated by <code>plink</code> into the format as <code>data/extra/haplotype_block_list.csv</code>. <br>
    ├── <code>0.3_Rice_HGRAINBOW_Simulation_of_phenotypic_values_and_some_preparation.R</code> : Simulate phenotypic values for both scenarios, coupling and repulsion. <br>
    ├── <code>0.4_Rice_HGRAINBOW_subpop_list_for_Table_S1.R</code> : Scripts for generating <code>Table_S1_origin.csv</code>. <br>
    ├── <code>1.1_Rice_HGRAINBOW_score_SKAT_geneset.R</code> : The function to run <code>SKAT</code> as haplotype-based GWAS method. <br>
    ├── <code>1.2_Rice_HGRAINBOW_haplotype_group_fixed_GWAS.R</code> : The function to run <code>HGF</code> as haplotype-based GWAS method. <br>
    ├── <code>1.3_Rice_HGRAINBOW_SS_gwas.R</code> : The function to calculate summary statistics from GWAS results. <br>
    ├── <code>2.1_Rice_HGRAINBOW_Haplotype_based_GWAS_for_RAINBOW_package_paper.R</code> : Perform each GWAS method and save the results. <br>
    ├── <code>3.1_Rice_HGRAINBOW_Summary_of_HGRAINBOW_Haplotype_based_GWAS_for_RAINBOW.R</code> : Summary GWAS results. <br>
    ├── <code>3.2_Rice_HGRAINBOW_Manhattan_plot_from_the_results.R</code> : Draw Manhattan plots from GWAS results. <br>
    ├── <code>3.3_Rice_HGRAINBOW_Summary_plot_of_HGRAINBOW_Haplotype_based_GWAS_for_RAINBOW.R</code> : Draw figures in the article from GWAS results. <br>
    └── <code>3.4_Rice_HGRAINBOW_Manhattan_plot_for_overwhelming_results.R</code> : Draw Manhattan plots for overwhelming results (<code>Figure 4</code>).</p></div></body></html>